examples/nn.dx

```diff
@@ -0,0 +1,227 @@
+    init = (\key.
+         [k1, k2] = splitKey key
+         (init ndense1 k1, init ndense2 k2))
+  }
+
+(all_params, fparams) = runState (init simple $ newKey 0) $ \params .
+  for _ : (Fin 500).
+       (loss, gradfn) =  vjp (\ params.
+                     -sum for j.
+                         result = forward simple params xs.j
+                         result.((y.j)@(Fin 2))) (get params)
+       gparams = gradfn 1.0
+       params := (get params) - scaleVec (0.05 / (IToF 100))  gparams
+       get params
+
+span = linspace (Fin 10) (-1.0) (1.0)
+tests = for h : (Fin 50). for i . for j.
+        r = forward simple all_params.((ordinal h * 10)@_) [span.i, span.j]
+        [exp r.(1@_), exp r.(0@_), 0.0]
+        
+:t tests
+
+:html imseqshow tests
+
+' ## LeNet for image classification
+
+H = 28
+W = 28
+Image = Fin H => Fin W => Float 
+Class = Fin 10
+
+
+lenet =
+  ncnn1 = cnn H W (Fin 1) (Fin 4) 3 3
+  ncnn2 = cnn H W (Fin 4) (Fin 8) 3 3
+  ndense1 = dense (Fin 392) (Fin 64)
+  ndense2 = dense (Fin 64) Class
+  {
+    forward = (\ param x.
+         (cnn1, cnn2, dense1, dense2) = param
+         x1' = forward ncnn1 cnn1 x
+         x1 = for i. for j. for k. relu x1'.i.j.k
+         x2' = forward ncnn2 cnn2 x1
+         x2 = for i. for j. for k. relu x2'.i.j.k
+         x3 : (Fin 8 => Fin 7 => Fin 7 => Float) = for c. meanpool (Fin 4) (Fin 4) x2.c
+         x3' = castTable (Fin 392) (for (i, j, k). x3.i.j.k)
+         x4' = forward ndense1 dense1 x3'
+         x4 = for i. relu x4'.i
+         logsoftmax $ forward ndense2 dense2 x4),
+    init = (\key.
+         [k1, k2, k3, k4] = splitKey key
+         (init ncnn1 k1, init ncnn2 k2,
+         init ndense1 k3, init ndense2 k4))
+  }
+
+
+
+
+' ## Data Loading
+
+  
+Batch = Fin 5000
+Full = Fin ((size Batch) * H * W)
+
+raw =
+    ls = unsafeIO $ \ _. readFile "examples/mnist.bin"
+    (AsList _ im) = ls
+    unsafeCastTable Full im
+
+def pixel (x:Char) : Float32 =
+     r = W8ToI x
+     IToF case r < 0 of
+             True -> (abs r) + 128
+             False -> r
+labels' =
+    ls = unsafeIO $ \ _. readFile "examples/labels.bin"
+    (AsList _ im) = ls
+    r = unsafeCastTable Batch im
+    for i. W8ToI r.i
+
+
+ims' =
+   for b: Batch.
+     for i:(Fin W).
+       for j:(Fin H).
+          pixel raw.((ordinal (b, i, j)) @ Full)
+
+
+' ## Training loop
+
+Epochs = (Fin 1)
+Minibatches = (Fin 250)
+Minibatch = (Fin 20)
+ims : Minibatches=>Minibatch=>Image = split ims'
+labels : Minibatches=>Minibatch=>Int = split labels'
+
+
+(losses, params) = runState (init lenet $ newKey 0) $ \params .
+  for _ : Epochs.
+    for i : Minibatches.
+       (loss, gradfn) =  vjp (\ params.
+                     -sum for j : Minibatch.
+                         result = forward lenet params (for c. ims.i.j)
+                         result.((labels.i.j)@Class)) (get params)
+       gparams = gradfn 1.0
+       params := (get params) - scaleVec (0.01 / (IToF (size Minibatch)))  gparams
+       loss
```

Comment:
Could you abstract this into a `def train` function? It shouldn't be too hard, right?

We have a training loop API now! We can consider exciting designs like [fast.ai's Learner](https://docs.fast.ai/learner.html), which is based on callbacks and hooks for e.g. learning rate scheduling, printing and recording metrics, etc.

Reply:
Let's slow down :) It will take a while until Dex will be viable for standard ML model training (mostly performance-wise), so this is perfectly fine.

Reply:
Haha yep. Not trying to speed up, just mentioning potential future ideas - in case others resonate or have different ideas.

Reply:
I can definitely abstract this loop. That's a good idea. I was mostly just too lazy to type it, so it was easier to write it twice.

Reply:
Mostly fixed here.  Not happy with minibatching if you have any ideas.

https://github.com/google-research/dex-lang/pull/456/files#diff-bcced3748a9b282a26c9bd64cd2a907277e61fbe272d8a7f4854c3f8c6fb4129R113

Reply:
Posting the minibatching code here for easier review:

```haskell
def train_class [VSpace p] (model: Layer a (b=>Float) p)
                           (x: batch=>a)
                           (y: batch=>b)
                           (epochs : Type)
                           (minibatch : Type)  (minibatches : Type) :
    (epochs => p & epochs => Float ) =
  xs : minibatches => minibatch => a = split x
  ys : minibatches => minibatch => b = split y
  ...
```

Reply:
Moving code review to the `train_class` function. Thanks!

Reply:
Right it feels like there should be a way to avoid passing in one of those arguments but not sure how to avoid it.